app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt

```diff
@@ -0,0 +1,280 @@
+        if (validation != null && validation.getStatus() !== ValidationResult.ValidationStatus.PARTIAL_NEXT && validation.getStatus() !== ValidationResult.ValidationStatus.VALID) {
+            Log.e(TAG, "Signed block did not validate. Result: " + validation.toString() + ". Errors: "
+                    + validation.getErrors().toString())
+        } else {
+            dbHelper.insertInDB(block)
+            return block
+        }
+        return null
+    }
+
+    fun sendCrawlRequest(peer: ChainService.Peer, publicKey: ByteArray, seqNum: Int): MutableIterator<ChainService.PeerTrustChainBlock> {
+        var sq = seqNum
+        if (seqNum == 0) {
+            sq = dbHelper.getBlock(publicKey,
+                    dbHelper.getMaxSeqNum(publicKey))?.sequenceNumber ?: TrustChainBlock.GENESIS_SEQ
+        }
+
+        if (sq >= 0) {
+            sq = Math.max(TrustChainBlock.GENESIS_SEQ, sq)
+        }
+
+        Log.i(TAG, "Requesting crawl of node " + Peer.bytesToHex(publicKey) + ":" + sq)
+
+        val crawlRequest = MessageProto.CrawlRequest.newBuilder()
+                .setPublicKey(ByteString.copyFrom(myPublicKey))
+                .setRequestedSequenceNumber(sq)
+                .setLimit(100).build()
+
+        // send the crawl request
+        val message = ChainService.PeerCrawlRequest.newBuilder().setPeer(me).setRequest(crawlRequest).build()
+        return registry.findStub(peer).recieveCrawlRequest(message)
+    }
+
+    override fun recieveCrawlRequest(crawlRequest: ChainService.PeerCrawlRequest, responseObserver: StreamObserver<ChainService.PeerTrustChainBlock>) {
+        val sq = sequenceNumberForCrawl(crawlRequest.request.requestedSequenceNumber)
+        println("processing crawl request")
+        Log.i(TAG, "Received crawl crawlRequest")
+
+        // a negative sequence number indicates that the requesting peer wants an offset of blocks
+        // starting with the last block
+        val blockList = dbHelper.crawl(myPublicKey, sq)
+
+        for (block in blockList) {
+            val peerTrustChainBlock = ChainService.PeerTrustChainBlock.newBuilder().setPeer(me).setBlock(block).build()
+            responseObserver.onNext(peerTrustChainBlock)
+        }
+        responseObserver.onCompleted()
+
+        Log.i(TAG, "Sent " + blockList.size + " blocks")
+    }
+
+    private fun sequenceNumberForCrawl(sq: Int): Int {
+        if (sq < 0) {
+            val lastBlock = dbHelper.getLatestBlock(myPublicKey)
+
+            if (lastBlock != null) {
+                return Math.max(TrustChainBlock.GENESIS_SEQ, lastBlock.sequenceNumber + sq + 1)
+            } else {
+                return TrustChainBlock.GENESIS_SEQ
+            }
+        } else {
+            return sq
+        }
+    }
+
+    companion object {
+        fun shouldSign(block: MessageProto.TrustChainBlock): Boolean {
+            return true
```

Comment:
Since we're going to have to make this part "interactive", should this be on the companion object?

Reply:
probably not, out of scope for this pr though